### PR TITLE
feat: Add `::fromGenerator` constructor.

### PR DESCRIPTION
**Pull request by drupol**
This PR:

* [x] Provide a new `Collection::fromGenerator(Generator $generator)` constructor.
* [x] Has unit tests (phpspec)
* [x] Has static analysis tests (psalm, phpstan)
* [x] Has documentation

### CODE REVIEW

File: docs/pages/api.rst [hunk 1]
```diff
@@ -48,14 +48,38 @@ Signature: ``Collection::fromIterable(string $filepath): Collection;``
...
+
 fromIterable
 ~~~~~~~~~~~~
 
 Create a collection from an iterable.
 
 .. warning:: When instantiating from a PHP `Generator`_, the collection object will inherit its behaviour:
     it will only be iterable a single time, and an exception will be thrown if multiple operations which attempt
-    to re-iterate are applied, for example ``count()``.
+    to re-iterate are applied, for example ``count()``. Use `Collection::fromGenerator()` to
```
Comment (AlexandruGG) [inline docs/pages/api.rst:81]:
```suggestion
    to re-iterate are applied, for example ``count()``. Use ``Collection::fromGenerator()`` to
```

File: docs/pages/api.rst [hunk 2]
```diff
@@ -48,14 +48,38 @@ Signature: ``Collection::fromIterable(string $filepath): Collection;``
...
     Collection::fromFile('http://loripsum.net/api');
 
+fromGenerator
+~~~~~~~~~~~~~
+
+Create a collection from a `Generator`_.
+
+.. warning:: The difference between this constructor and ``fromIterable`` is that
+    the generator is decorated with a caching Iterator. ``Generators`` are not
+    **rewindable** by design and using ``fromIterable`` automatically adds the
```
Comment (AlexandruGG) [inline docs/pages/api.rst:58]:
```suggestion
    **rewindable** by design and using ``fromGenerator`` automatically adds the
```

I think you meant this here?

File: docs/pages/api.rst [hunk 3]
```diff
@@ -48,14 +48,38 @@ Signature: ``Collection::fromIterable(string $filepath): Collection;``
...
+
+Signature: ``Collection::fromGenerator(Generator $generator): Collection;``
+
+.. code-block:: php
+
+    $generator = (static fn () => yield from range(1, 10))();
+    $generator->next();
+    $generator->next();
+
+    $collection = Collection::fromGenerator($generator);
```
Comment (AlexandruGG) [inline docs/pages/api.rst:72]: this example doesn't tell me much :(. Can we show how this collection will iterate? For example by applying `all` or using a `foreach` loop?

File: spec/loophp/collection/CollectionSpec.php [hunk 4]
```diff
@@ -305,9 +305,39 @@ public function it_can_be_constructed_from_a_generator(): void
...
+        };
+
+        $generator = $generator();
+        $generator->next();
+        $generator->next();
+
+        $subject = $this::fromGenerator($generator);
+
+        $subject
+            ->shouldIterateAs([2 => 'c', 3 => 'd', 4 => 'e']);
```
Comment (AlexandruGG) [inline spec/loophp/collection/CollectionSpec.php:336]: is it possible to have a test where we do the same thing but with `fromIterable` so we can see what the difference is (if any) ?

Comment (AlexandruGG) [reply spec/loophp/collection/CollectionSpec.php:336]: I'm guessing `fromIterable` will attempt to rewind the Generator and it fails actually, right?

File: docs/pages/api.rst [hunk 5]
```diff
@@ -48,14 +48,38 @@ Signature: ``Collection::fromIterable(string $filepath): Collection;``
...
 fromIterable
 ~~~~~~~~~~~~
 
 Create a collection from an iterable.
 
 .. warning:: When instantiating from a PHP `Generator`_, the collection object will inherit its behaviour:
     it will only be iterable a single time, and an exception will be thrown if multiple operations which attempt
-    to re-iterate are applied, for example ``count()``.
+    to re-iterate are applied, for example ``count()``. Use `Collection::fromGenerator()` to
+    circumvent this limitation.
```
Comment (AlexandruGG) [inline docs/pages/api.rst:82]: we could mention that another alternative is using `fromCallable` with a callable that returns the Generator

File: docs/pages/api.rst (hunk 2 above)
Comment (drupol) [reply docs/pages/api.rst]: Yes, thanks !

File: docs/pages/api.rst (hunk 5 above)
Comment (drupol) [reply docs/pages/api.rst]: Sure, it's done!

File: spec/loophp/collection/CollectionSpec.php (hunk 4 above)
Comment (drupol) [reply spec/loophp/collection/CollectionSpec.php:337]: Job done!

File: docs/pages/api.rst (hunk 3 above)
Comment (drupol) [reply docs/pages/api.rst]: Work complete! https://www.youtube.com/watch?v=PoMhMZZjbGg

**Review by AlexandruGG — APPROVED** [review]
Nice one!

5hozbx

**Pull request merged**